integration-tests/slack/slack-utils.js

```diff
@@ -0,0 +1,148 @@
+const fs = require('fs')
+const _ = require('lodash/fp')
+const fetch = require('node-fetch')
+const { parse } = require('path')
+const { postMessage } = require('./post-message')
+const { printMessageTemplate } = require('./message-templates')
+
+
+const {
+  SLACK_BOT_TOKEN: defaultToken,
+  JOB_BUILD_NUM: circleJobBuildNum
+} = process.env
+
+/**
+ * Fetch CircleCI artifact links to tests summary JSON files (were created in onRunComplete() in jest-reporter.js)
+ * @param { string } token
+ * @param { string } buildNum
+ * @returns { Promise<Array[string]> } URL to tests-summary-[0-9].json
+ */
+const fetchCircleJobArtifacts = async ({ token = defaultToken, buildNum = circleJobBuildNum } = {}) => {
+  if (!buildNum) {
+    throw new Error(`**  ERROR: Missing CircleCI build number. Failed to fetch CircleCI job artifacts.`)
+  } else if (!token) {
+    throw new Error(`**  ERROR: Missing CircleCI API token. Failed to fetch CircleCI job artifacts.`)
+  }
+
+  // Find more arguments and details of the response: https://circleci.com/docs/api/v2/index.html#operation/getJobArtifacts
+  const apiUrl = 'https://circleci.com/api/v2/project/github/DataBiosphere/terra-ui'
+  const headers = { Accept: 'application/json', Authorization: `Bearer ${token}` }
+
+  try {
+    const response = await fetch(`${apiUrl}/${buildNum}/artifacts`, { method: 'GET', headers })
+    const json = await response.json()
+    const { items } = json
+    const filtered = _.filter(item => item['path'].toString().includes('tests-summary'), items)
+    return filtered.map(item => item.url)
+  } catch (e) {
+    console.error(`**  ERROR: Encountered unexpected error when getting CircleCI build_number: ${buildNum} artifacts:`, e)
+    throw e
+  }
+}
+
+const getFailedTestFileName = aggregatedResult => {
+  const failedTests = []
+  _.forEach(testResult => {
+    if (testResult.numFailingTests > 0) {
+      failedTests.push(parse(testResult.testFilePath).name)
+    }
+  }, aggregatedResult.testResults)
+  return failedTests
```

Comment:
This can be written with `filter` and `map` instead of appending to an array.

```suggestion
  return _.flow(
    _.filter(testResult => testResult.numFailingTests > 0)
    _.map(testResult => parse(testResult.testFilePath).name)
  )(aggregatedResult.testResults)
```

https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide#coding-style

Reply:
Thanks.
I tried to wrtie FP functions as best as I can but I'm not an advanced JS programmer. I focused on making it work, I have not iterated towards better code styles. I'll try do better next time.
Just my own opinion, such guide is not really useful to me because it doesn't actually show how to guide someone like me to write better FP code. Real code examples will be helpful if they exist.

Reply:
I linked to that mainly for the "Use Lodash/fp for data transformation" guideline.

Leaving out `_.flow` and Lodash entirely, this could also have been written as:
```js
return aggregatedResult.testResults
  .filter(testResult => testResult.numFailingTests > 0)
  .map(testResult => parse(testResult.testFilePath).name)
```

Generally, I think the functional style and immutable data helps reason about what code is doing by more clearly showing how data is being transformed. As opposed to mutable state (such as the `failedTests` array in the previous version of this code) where you have to step through how it's changing over time to understand how the output relates to the input.